### DISCUSSION TITLE
Business  Rules in BusinessListBase

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3215] [Questions]
We are having fun with a Csla based project from a few years ago.
I noticed that in order to avoid duplicate items in a collection (BusinessListBase) we check the contains method first and then add the item to the collection.

Since BusinessListBase doesn't support 'AddBusinessRules (?) we can't use something like NoDuplicateResource in ProjectTracker. I am concerned that UI developers (in the future) might not check the contains method, is there a way we could write a validation rule?

Kind regards

**Comment by JacoJordaan** [comment]
Here is an example of how I do this with a property rule:
````
    internal sealed class PaymentMethodAlreadyExistsInListRule : PropertyRule
    {
        internal PaymentMethodAlreadyExistsInListRule(IPropertyInfo primaryProperty)
          : base(primaryProperty)
        {
            if (InputProperties == null)
                InputProperties = new List<IPropertyInfo>();
            InputProperties.Add(primaryProperty);

            IsAsync = false;
            CanRunAsAffectedProperty = true;   // only run when this property is changed
            CanRunInCheckRules = true;          // when true will also run in CheckRules
            CanRunOnServer = false;             // when true will also run on logical serverside (Data Access)
        }

        protected override void Execute(IRuleContext context)
        {           
            var bo = (PaymentSetupPaymentMethodEdit)context.Target;
            var paymentMethodInput = (string)context.InputPropertyValues[PrimaryProperty];

            if (bo != null)
            {
                var paymentMethods = (PaymentSetupPaymentMethods)bo.Parent;

                if (paymentMethods != null)
                {
                    foreach (var paymentMethod in paymentMethods)
                    {
                        if (paymentMethod != bo)
                        {
                            if (paymentMethod.PaymentMethod == paymentMethodInput)
                            {
                                context.AddErrorResult("Payment method already exists in list of payment methods.");
                                break;
                            }
                        }
                    }
                }
            }
        }
    }
````

**Reply by Chicagoan2016** [reply]
@JacoJordaan I would have never thought of implementing something like this 👍
We do have unique index in the database table. I remember reading in Using Csla book for Mvc that sometimes we have to rely  on database(only) for referencial integrity.

**Comment by rockfordlhotka** [comment]
The suggested answer at the moment is to have a parent `BusinessBase` type that contains the list, then you can have rules in the parent - though you still need to trigger them by handling the `ChildChanged` event to know when something has changed within the list.

There's been talk about a (major) change to the framework to merge `BusinessBase` and `BusinessListBase` into a single type - thus allowing the rules engine and properties to all be used in a list type. That wouldn't work with Windows Forms or Web Forms, but modern UI frameworks for XAML and Blazor would handle the concept just fine.
